### PR TITLE
Update openjpeg 2.3

### PR DESCRIPTION
**Pull request by pierreldff**

### CODE REVIEW

File: Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.h [hunk 1]
```diff
@@ -81,6 +85,7 @@ friend class ImageRegionReader;
   bool CodeFrameIntoBuffer(char * outdata, size_t outlen, size_t & complen, const char * indata, size_t inlen );
   bool GetHeaderInfo(const char * dummy_buffer, size_t len, TransferSyntax &ts);
   JPEG2000Internals *Internals;
+  int nNumberOfThreads;
```
Comment (malaterre) [inline Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.h]: pimpl please...

Comment (pierreldff) [reply Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.h]: Ok no problem

File: Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.h [hunk 2]
```diff
@@ -56,6 +56,10 @@ friend class ImageRegionReader;
 
   void SetNumberOfResolutions(unsigned int nres);
 
+  /// Set Number of threads
+  /// @param nThreads : number of threads for codec, if negative value is set determine how many virtual threads are available
+  void SetNumberOfThreads(int nThreads);
```
Comment (malaterre) [inline Source/MediaStorageAndFileFormat/gdcmJPEG2000Codec.h]: @pierreldff Can you update the documentation for the case `0`, it should be clear to the user that this means current thread (no new thread created if I understand correctly).

**Comment by issakomi** [comment]
@malaterre 
looks like
gdcmconv -K intput.dcm out.dcm
crashes highly likely on multi-threading stuff, debugger points to (GDCM configured with default settings)
could you reproduce? (i think with any image)

**Comment by issakomi** [comment]
@malaterre 
related to VR changes (not OpenJPEG, but merged with this commit) - they are very very good, but will not work on Windows, i have done it already yesterday, shall make PR with fix, tested read/write all

Actually with VS 2013 x64 can not compile

`12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(84): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(93): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToEncoding<0>' has already been defined
12>          c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToEncoding<0>'
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToType<0>' has already been defined
12>          c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToType<0>'
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(356): error C2196: case value '0' already used
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(360): error C2196: case value '0' already used
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(369): error C2196: case value '1' already used
12>  gdcmDataSet.cxx
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(84): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(93): warning C4309: 'initializing' : truncation of constant value
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToEncoding<0>' has already been defined
12>          c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToEncoding<0>'
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(313): error C2766: explicit specialization; 'gdcm::VRToType<0>' has already been defined
12>          c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(304) : see previous definition of 'VRToType<0>'
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(356): error C2196: case value '0' already used
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(360): error C2196: case value '0' already used
12>c:\users\mi\desktop\gdcm\source\datastructureandencodingdefinition\gdcmVR.h(369): error C2196: case value '1' already used
12>  gdcmExplicitDataElement.cxx`

**Comment by malaterre** [comment]
@pierreldff I assumed you had tested those changes. Indeed I can reproduce the segfault overhere. Looks like the travis/appveyor are not setup to do proper testing for this case. I'll have to backout your changes unless you can provide a fix quickly.

@issakomi sorry for the long vs int truncation issue in gdcm::VR, so you think you can provide a pull request in the upcoming days ?

**Comment by pierreldff** [comment]
Yes I have test on windows and i compare time between one thread and multithread with gdcmconv
I have test a decompression of big files

**Comment by malaterre** [comment]
@pierreldff I can reproduce the segfault (NULL ptr) here,  @issakomi can also, and all MacOSX dashboard fails the JPEG2000 tests:

* https://open.cdash.org/testSummary.php?project=21&name=TestImageChangeTransferSyntax2&date=2019-04-05
* https://open.cdash.org/testSummary.php?project=21&name=TestFileChangeTransferSyntax3&date=2019-04-05

I'll give it one quick try today, if I cannot make sense of the segfault I'll revert. We'll have the history here anyway. Thanks

**Comment by pierreldff** [comment]
Ok there were a problem with compression that do not support thread i have done an other pull request.
Sorry for that

**Comment by pierreldff** [comment]
I test with decompression https://github.com/malaterre/GDCM/pull/87

**Comment by issakomi** [comment]
Were good to update the version in [cmake file](https://github.com/malaterre/GDCM/blob/e2dbdbeca6402278a2a8eb36eeeed51d95c78662/Utilities/gdcmopenjpeg/CMakeLists.txt#L45) (may be also review the file, it is from previous distribution?) and [README.GDCM.txt](https://github.com/malaterre/GDCM/blob/master/Utilities/gdcmopenjpeg/README.GDCM.txt)
Many thanks.

**Comment by pierreldff** [comment]
Sorry,
I'm at the connectathon this week. I'll update those files next week.

**Comment by issakomi** [comment]
#88

**Comment by issakomi** [comment]
Updated openjpeg_mangle.h.in
#89

**Pull request merged**